retrieval/targetmanager.go

```diff
@@ -14,139 +14,356 @@
+	targets map[string][]Target
+	// Providers and configs by their job name.
+	// TODO(fabxc): turn this into map[*ScrapeConfig][]TargetProvider eventually.
+	providers map[string][]TargetProvider
+	configs   map[string]config.JobConfig
 }
 
-// NewTargetManager returns a newly initialized TargetManager ready to use.
-func NewTargetManager(sampleAppender storage.SampleAppender, globalLabels clientmodel.LabelSet) TargetManager {
-	return &targetManager{
+// NewTargetManager creates a new TargetManager based on the given config.
+func NewTargetManager(cfg config.Config, sampleAppender storage.SampleAppender) (*TargetManager, error) {
+	tm := &TargetManager{
 		sampleAppender: sampleAppender,
-		globalLabels:   globalLabels,
-		poolsByJob:     make(map[string]*TargetPool),
+		targets:        make(map[string][]Target),
+	}
+	if err := tm.updateConfig(cfg); err != nil {
+		return nil, err
 	}
+	return tm, nil
 }
 
-func (m *targetManager) targetPoolForJob(job config.JobConfig) *TargetPool {
-	targetPool, ok := m.poolsByJob[job.GetName()]
+// Run starts background processing to handle target updates.
+func (tm *TargetManager) Run() {
+	glog.Info("Starting target manager...")
 
-	if !ok {
-		var provider TargetProvider
-		if job.SdName != nil {
-			provider = NewSdTargetProvider(job, m.globalLabels)
-		}
+	sources := make(map[string]bool)
+
+	for name, provs := range tm.providers {
+		for _, p := range provs {
+			jcfg := tm.configs[name]
 
-		interval := job.ScrapeInterval()
-		targetPool = NewTargetPool(provider, m.sampleAppender, interval)
-		glog.Infof("Pool for job %s does not exist; creating and starting...", job.GetName())
+			ch := make(chan *config.TargetGroup)
+			go tm.run(tm.configs[name], ch)
 
-		m.poolsByJob[job.GetName()] = targetPool
-		go targetPool.Run()
+			for _, src := range p.Sources() {
+				src = fullSource(jcfg, src)
+				sources[src] = true
+			}
+
+			// Run the target provider after cleanup of the stale targets is done.
+			defer func(c chan *config.TargetGroup) {
+				go p.Run(c)
+			}(ch)
+		}
 	}
 
-	return targetPool
+	tm.removeTargets(func(src string) bool {
+		return !sources[src]
+	})
+
+	tm.running = true
 }
 
-func (m *targetManager) AddTarget(job config.JobConfig, t Target) {
-	m.Lock()
-	defer m.Unlock()
+// run receives target group updates and handles them in the context of the
+// given job config.
+func (tm *TargetManager) run(cfg config.JobConfig, ch <-chan *config.TargetGroup) {
+	for {
+		tg, open := <-ch
+		if !open {
+			break
+		}
+		glog.V(1).Infof("Received update for target group %q", tg.Source)
 
-	targetPool := m.targetPoolForJob(job)
-	targetPool.AddTarget(t)
-	m.poolsByJob[job.GetName()] = targetPool
+		if err := tm.updateTargetGroup(tg, cfg); err != nil {
+			glog.Errorf("Error updating targets: %s", err)
+		}
+	}
 }
 
-func (m *targetManager) ReplaceTargets(job config.JobConfig, newTargets []Target) {
-	m.Lock()
-	defer m.Unlock()
+// fullSource prepends the source by the unique job name.
```

Comment:
Not sure if `prepends by` is correct grammar. `prepends the unique job name to the source.`?